bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -4099,6 +4110,16 @@ ReplicaImp::ReplicaImp(bool firstTime,
       KeyExchangeManager::instance().sendInitialKey(currentPrimary());
     }
   });
+  stateTransfer->addOnFetchingStateChangeCallback([&](uint64_t checkPointNum) {
+    // With (n-f) initial key exchange support, if we have a lagged replica
+    // which syncs its state through ST, we need to make sure that it completes
+    // initial key exchange after completing ST
+    if (!isCollectingState()) {
+      if (ReplicaConfig::instance().getkeyExchangeOnStart() && !KeyExchangeManager::instance().exchanged()) {
+        KeyExchangeManager::instance().sendInitialKey(currentPrimary(), checkPointNum * checkpointWindowSize);
```

Comment:
Nit: You can use here the `lastExecutedSequenceNum` or `lastStableSeqNum` instead of `checkPointNum * checkpointWindowSize`